test/index.js

```diff
@@ -1,55 +1,339 @@
-var apply = require('../apply')
-
-var hello = require('../examples/hello')
 var Combine = require('../')
-var tape = require('tape')
+var test = require('tape')
+
+test('combine is a function', function (t) {
+  t.equal(typeof Combine, 'function')
+  t.end()
+})
 
-tape('simple', function (t) {
-  t.equal(Combine([hello.hi, hello.greet]).hello[0]('depject'), 'Hello, depject')
-  t.equal(Combine([hello.hi, hello.capitalize, hello.greet]).hello[0]('depject'), 'Hello, DEPJECT')
 
+test('combine one module', function (t) {
+  const cats = {
+    gives: 'cats',
+    create: () => {
+      t.ok(true, 'create is called')
+      return () => () => true
+    }
+  }
+
+  var api = Combine([cats])
+
+  t.ok(api.cats, 'Combine returns an object with keys that match the keys given by all the modules')
   t.end()
 })
 
-tape('entry', function (t) {
-  console.log(Combine([hello.hi, hello.greet]).hello)
-  t.equal(apply.first(Combine([hello.hi, hello.greet]).hello)('depject'), 'Hello, depject')
-  t.equal(apply.first(Combine([hello.hi, hello.capitalize, hello.greet]).hello)('depject'), 'Hello, DEPJECT')
+test('combine two modules', function (t) {
+  const cats = {
+    gives: 'cats',
+    create: () => {
+      t.ok(true, 'create is called')
+      return () => () => true
+    }
+  }
+
+  const client = {
+    needs: {cats: 'first'},
+    gives: 'client',
+    create: (api) => {
+      t.ok(api.cats, 'create is called with object that has keys given by other modules')
+      return () => () => true
+    }
+  }
+
+  var api = Combine([client, cats])
 
+  t.ok(api.cats && api.client, 'Combine returns an object with keys that match the keys given by all the modules')
   t.end()
 })
 
+test('combine takes an array of modules and throws if not an array', function (t) {
+  t.plan(1)
 
-var isModule = require('../is')
+  const cats = {
+    gives: 'cats',
+    create: () => {
+      return () => () => true
+    }
+  }
 
-tape('isModule', function (t) {
+  t.throws(() => Combine(cats))
+})
+
+test('one module depends on a module that depends on another', function (t) {
+  const a = {
+    needs: {b: 'first'},
+    gives: 'a',
+    create: (api) => api.b
+  }
+  const b = {
+    needs: {c: 'first'},
+    gives: 'b',
+    create: (api) => api.c
+  }
+  const c = {
+    gives: 'c',
+    create: () => () => true
+  }
 
-  t.notOk(isModule(hello))
-  for(var k in hello)
-    t.ok(isModule(hello[k]))
+  var api = Combine([a, b, c])
+  t.ok(api.a[0]())
   t.end()
+})
+
+test('two modules depend on each other', function (t) {
+  const cats = {
+    needs: {dogs: 'first'},
+    gives: 'cats',
+    create: (api) => {
+      t.ok(api.dogs, 'api provides the needed dep')
```

Comment:
maybe test that it's not resolved immediately (will throw), but is resolved after a `process.nextTick`?